### PR TITLE
Compute Switch Threshold

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Missing computation of switch threshold for optimistic confirmation

#### Summary of Changes
1) Refactor ReplayStage to not vote when switching threshold fails, instead reset to heaviest descendant of last vote (Will factor this out into another PR)

2) Compute switching threshold 

Fixes #

### CODE REVIEW

File: core/src/consensus.rs [hunk 1]
```diff
@@ -319,6 +336,88 @@ impl Tower {
...
+                    // Find any locked out intervals in this bank with endpoint >= last_vote,
+                    // implies they are locked out at last_vote
+                    for (_, value) in lockout_intervals.range((Included(last_vote), Unbounded)) {
+                        for (lockout_interval_start, vote_account_pubkey) in value {
+                            // Only count lockouts on slots that are:
+                            // 1) Not ancestors of `last_vote`
+                            // 2) Not from before the current root as we can't determine if
+                            // anything before the root was an ancestor of `last_vote` or not
+                            if !last_vote_ancestors.contains(lockout_interval_start)
+                                && ancestors.contains_key(lockout_interval_start)
```
Comment (carllin) [inline core/src/consensus.rs:403]: @aeyakovenko, I'm filtering out any branches from before the root, so those forks can't be included in the switching proofs, even though they may be locked out above our last vote. I don't think it's a huge issue, but I could be missing something...

File: core/src/consensus.rs [hunk 2]
```diff
@@ -115,6 +120,13 @@ impl Tower {
             }
             let mut vote_state = vote_state.unwrap();
 
+            for vote in &vote_state.votes {
+                lockout_intervals
+                    .entry(vote.expiration_slot())
+                    .or_insert_with(|| vec![])
+                    .push((vote.slot, key));
```
Comment (carllin) [inline core/src/consensus.rs:127]: Will add these keys to the replay_stage `all_pubkeys` so they pull from the same reference pool to reduce memory usage.

File: core/src/consensus.rs [hunk 3]
```diff
@@ -319,6 +336,91 @@ impl Tower {
...
+                                    .unwrap_or(0);
+                                locked_out_stake += stake;
+                                locked_out_vote_accounts.insert(vote_account_pubkey);
+                            }
+                        }
+                    }
+                }
+                (locked_out_stake as f64 / total_stake as f64) > SWITCH_FORK_THRESHOLD
+            })
+            .unwrap_or(true)
```
Comment (carllin) [inline core/src/consensus.rs:421]: This doesn't generate the proof yet, want to make sure these incremental changes don't break anything first

**Comment by aeyakovenko** [comment]
Looks pretty good.  We really need to pull the consensus stuff out into a non networked simulation environment.

**Comment by carllin** [comment]
Seems to be stable for last 5 hours with regular partitioning every 10 mins
Screen Shot 2020-05-11 at 12 13 21 PM

**Comment by aeyakovenko** [comment]
@carllin do these tests verify that the network can recover from a 33 - 4 failure?  Assuming the threshold is 4

**Comment by carllin** [comment]
@aeyakovenko, if the threshold is 4, then the switching threshold is `(100 + 4 - 66.66) = 37.34`. Then you can tolerate at most `100 - 37.4 * 2 = 25.2` failure. Any more than 25% that and assuming the worst case 50-50 partition, then validators will get stuck on their respective partitions without being able to switch.

There's currently no test that sets up this exact scenario, but this one is pretty close: https://github.com/solana-labs/solana/blob/master/local-cluster/tests/local_cluster.rs#L376. It currently kill `9/29 ~= 31%` of the stake, but I think it currently passes because it kills the validator *after* the partition is resolved: https://github.com/solana-labs/solana/blob/master/local-cluster/tests/local_cluster.rs#L319-L324, which I think means the validator has voted sufficiently during/after the partition to use their votes in a switching proof.

 If we were instead to toggle the test to kill the leader before the partition, that would probably test this case.

**Comment by aeyakovenko** [comment]
@carllin but once the partition recovers the nodes can come back.  ideally we have a local cluster test and a nightly partition test that induces this scenario.

**Comment by carllin** [comment]
@aeyakovenko even after the partition resolves, if > 25% are dead then each side of the partition may get stuck, as there's not enough stake on the other side of the partition to generate a switching proof. 

The only possible way out of this hole is if then if the smaller/less staked partition then *itself* sub-partitions/forks, and then people on that side of the partition also vote on their own fork (they would have to think their fork is the heaviest, which may not happen if they detect the major fork)  allowing some of the validators on that side of the partition to generate a switching proof to switch. But this seems like a very unlikely escape

We can add a local cluster test, the nightly partition test can be an expansion of the existing nightly partition tests + ability to kill some of the nodes.

**Comment by aeyakovenko** [comment]
@carllin i meant the network should recover as soon as we are under 25% dead

**Pull request merged**